text/3502-cargo-script.md

```diff
@@ -0,0 +1,1229 @@
+- For `Cargo.toml`, `package.workspace = true` is the default
+- For single-file packages, `package.workspace = false` is the default
+
+When a workspace is specified
+- Use its target directory
+- Use its lock file
+- Be treated as any other workspace member for `cargo <cmd> --workspace`
+- Check what `workspace.package` fields exist and automatically apply them over default manifest fields
+- Explicitly require `workspace.dependencies` to be inherited
+  - I would be tempted to auto-inherit them but then `cargo rm`s gc will remove them because there is no way to know they are in use
+- Apply all `profile` and `patch` settings
+
+This could serve as an alternative to
+[`cargo xtask`](https://github.com/matklad/cargo-xtask) with scripts sharing
+the lockfile and `target/` directory.
+
+## Scaling up
+
+We provide a workflow for turning a single-file package into a multi-file
+package, on `cargo-new` / `cargo-init`.  This would help smooth out the
+transition when their program has outgrown being in a single-file.
+
+## A REPL
+
+See the [REPL exploration](https://github.com/epage/cargo-script-mvs/discussions/102)
+
+In terms of the CLI side of this, we could name this `cargo shell` where it
+drops you into an interactive shell within your current package, loading the
+existing dependencies (including dev).  This would then be a natural fit to also have a `--eval
+<expr>` flag.
+
+Ideally, this repl would also allow the equivalent of `python -i <file>`, not
+to run existing code but to make a specific file's API items available for use
+to do interactive whitebox testing of private code within a larger project.
+
+## Embedded or adjacent Lockfile
+
+[Lockfiles](https://doc.rust-lang.org/cargo/guide/cargo-toml-vs-cargo-lock.html)
+record the exact version used for every possible dependency to ensure
+reproducibility.  In particular, this protects against upgrading to broken
+versions and allows continued use of a yanked version.  As this time, the
+recommendation is for
+[`bin`s to persist their lockfile while `lib`s do not](https://doc.rust-lang.org/cargo/faq.html#why-do-binaries-have-cargolock-in-version-control-but-not-libraries).
+
+With multi-file packages, `cargo` writes a `Cargo.lock` file to the package
+directory.  As there is no package directory for single-file packages, we need
+to decide how to handle locking dependencies.
+
+Considerations
+- Sharing of single-file projects should be easy
+  - In "copy/paste" scenarios, like reproduction cases in issues, how often
+    have lockfiles been pertinent for reproduction?
+- There is an expectation of a reproducible Rust experience across systems with lockfiles being a key element
+- Dropping of additional files might be frustrating for users to deal with (in
+  addition to making it harder to share it all)
+- We would need to decide what to do for lockfiles and `stdin` (another future possibility) without conflicting
+  with parallel runs
```

Comment:
the current plan is apparently to use a hash of the file path, but that would also be difficult for `stdin`. would it make sense to hash the file content instead?

Reply:
Yes, that is a possibility  As this is a future possibility, the point is more to highlight the idea for us to include in consideration.  I don't think we need to work out all the details at this time.